astropy/units/utils.py

```diff
@@ -214,7 +214,7 @@ def maybe_simple_fraction(p, max_denominator=100):
     """
     if p == 0 or p.__class__ is int or p.__class__ is Fraction:
         return p
-    n, d = p.as_integer_ratio()
```

Comment:
Comment to the if above, which is not new code:

the python data model API for "integer-like" things is `__index__`, so you could check for:

```
if hasattr(p, "__index__"):
    return p.__index__()
```

https://docs.python.org/3/reference/datamodel.html#object.__index__

Reply:
that e.g. works for the numpy integer types which currently fail. This is of course fixed with your addition of `float(p)`, but then you run the algorithm needlessly for integers.

Reply:
I think there may be two different concerns here:
- fixing a regression
- making sure performances are optimal

In my opinion it's best to keep this PR focused on the first goal, but we can certainly tackle the second one in a follow up PR !
In the meantime, if anyone is worried about performance regressions from my proposed patch, we could run benchmarks on this PR

Reply:
I added `benchmark`  label. FYI.